WIP studies and fixes with generate_events

valassi opened the pull request:
This is a WIP PR with fixes and studies in generate_events

It includes a possible fix for #947

valassi commented:
As suggested by @oliviermattelaer the fix for #947 is moved to https://github.com/mg5amcnlo/mg5amcnlo/pull/125

valassi commented:
The fix for #947 has been stripped off in #949 (which can be merged now)

This has been rebased over it and remains in WIP

valassi commented:
This is now including quite a few other things, including patches for 
- #945 produce multi-build gridpacks (this is the default in my lauX.sh script: whenever a generate_events is tested, the code is in any case built for all backends; when a backend is specified to produce a gridpack, that backend is used for the actual vegas optimisation, but the code is built anyway and included in tar.gz for all backends; when the -ALL option is specified with -fromgridpack, then all backends are tested one after the other, with the only minor modifiction that the madevent symlink is moved in the untarred gridpack)
- (note that the above makes #674 a bit less relevant/urgent, there is no need to add a switche script for madevent, as effectivley redefining symlinks provides the backend switch)
- #957 instrument python code to provide time profiles
- (this PR internally includes the clang format PR #953 because I needed this to use itgold91 to regenerate code, while I was having issues on itscrd90)

Note that this PR is then internally included by the CMS DY studies PR #946. All of the goodies described above are used to provide a time profile on CMS DY and hel debug the CMS speedup issues #943 reported by @choij1589 )

An example of an instrumented file on gqttq is here https://github.com/madgraph5/madgraph4gpu/blob/1cbca07e243ade04c612cabb2cc253bc17e81a9c/epochX/cudacpp/tlau/fromgridpacks/gq_ttq.mad/cppnone/output.txt

valassi commented:
I have resynced this to upstream/master and fixed conflicts

valassi commented:
OK current status for timer/counter, grid/runcard, cmsdy and new sampling branches:
- cmsdyps branch in #970
- includes cmsdy branch #946 
- which includes THIS grid branch #948 
- which includes the current master
- [conversely, prof branch #962 includes master, but is still not fully synced with the rest]

valassi commented:
(There have been some updates in many branches: in THIS grid I regenerated all processes)

Updated status for timer/counter, grid/runcard, cmsdy and new sampling branches:
- cmsdyps branch is #970
- cmsdyps includes cmsdy branch #946 
- cmsdy includes THIS grid branch #948 which includes the current master
- cmsdy also includes prof branch #962 which includes the current master
- currently all processes are regenerated in each branch: cmsdy and cmsdyps also include dy012j and dy3j

valassi commented:
I have updated branch runcard with upstream/master.
And then updated this branch grid with the updated runcard.
So there are no conflicts with master here.

I am now running a DY+3jet gridpack creation and execution test.

oliviermattelaer commented:
Hi Andrea,

I would propose to first focus on merging for360 branch (and therefore goodhel). Such that we can reduce the number of open branch PR.

valassi commented:
> Hi Andrea,
> 
> I would propose to first focus on merging for360 branch (and therefore goodhel). Such that we can reduce the number of open branch PR.

Hi Olivier, thanks. I agree, when I am back from holidays on Thursday I will look at that. Tonight I just had bits and pieces of time in my spare time and I thought that this may be useful for CMS.

@choij1589 (and @Saptaparna) if you really want to try this AS-IS no guarantees (and knowing that it will take 1-2 weeks before this gets merged in production - and knowing that the user interfaces will probably change), you can try my branch https://github.com/valassi/madgraph4gpu/tree/grid. For instance at this commit
```
git clone --recursive https://github.com/valassi/madgraph4gpu.git
cd madgraph4gpu
git checkout grid
git reset --hard bad260f
git submodule update
```

In your runcard you must change False to True (or add this but I think you will see it) for this card
```
 True = cudacpp_bldall ! CUDACPP build all available backends in separate build directories: False, True
```
and try to build gridpacks using cuda
```
 True     = gridpack  !True = setting up the grid pack
 cuda = cudacpp_backend ! CUDACPP backend: fortran, cuda, hip, cpp, cppnone, cppsse4, cppavx2, cpp512y, cpp512z, cppauto
```

The above is for creating the multi-backend gridpack. Then in order to execute it you will simply need to redefine symlinks. This is not in any user interfaces yet, but you can see how I did it in my script below, just to give you an idea (or ask me)
https://github.com/valassi/madgraph4gpu/blob/bad260ff6cc26c802a006c87618a1399824f58ad/epochX/cudacpp/tlau/lauX.sh#L173
```

  # Configure the appropriate backend
  for dir in madevent/SubProcesses/P*; do
    pushd $dir >& /dev/null
    echo "INFO: redefine madevent symlink for backend ${bckend} in ${dir}"
    if [ "${bckend}" == "fortran" ]; then
      exe=$(\ls madevent_fortran)
    else
      exe=$(\ls build.*${bckend/cpp/}*/madevent_*)
    fi
    if [ "$(echo ${exe} | wc -w)" == "0" ]; then
      echo "ERROR! No madevent executable found for backend $bckend"; exit 1
    elif [ "$(echo ${exe} | wc -w)" != "1" ]; then
      echo "ERROR! Too many madevent executables found for backend $bckend:"; echo "$exe"; exit 1
    fi
    \rm -f madevent
    ln -sf $exe madevent
    echo "----> madevent symlink will now point to $exe"
    popd >& /dev/null
  done
```

I have done some tests and the results do seem to have event generation faster in AVX2 than in fortran for DY+3j (I did not try inclusive 0123j). Also these results are very preliminary and to be taken with a pinch of salt. The cuda result is weird as the non-ME part seems slower than in fortran/cpp. See https://github.com/valassi/madgraph4gpu/blob/bad260ff6cc26c802a006c87618a1399824f58ad/epochX/cudacpp/tlau/fromgridpacks/pp_dy3j.mad/summary.txt
```
pp_dy3j.mad/fortran/output.txt
[GridPackCmd.launch] OVERALL TOTAL    515.6369 seconds
[madevent COUNTERS]  PROGRAM TOTAL    498.905
[madevent COUNTERS]  Fortran Overhead 168.037
[madevent COUNTERS]  Fortran MEs      330.869
--------------------------------------------------------------------------------
pp_dy3j.mad/cppnone/output.txt
[GridPackCmd.launch] OVERALL TOTAL    504.0330 seconds
[madevent COUNTERS]  PROGRAM TOTAL    486.688
[madevent COUNTERS]  Fortran Overhead 165.031
[madevent COUNTERS]  CudaCpp MEs      320.487
[madevent COUNTERS]  CudaCpp HEL      1.1708
--------------------------------------------------------------------------------
pp_dy3j.mad/cppsse4/output.txt
[GridPackCmd.launch] OVERALL TOTAL    350.8207 seconds
[madevent COUNTERS]  PROGRAM TOTAL    334.902
[madevent COUNTERS]  Fortran Overhead 165.453
[madevent COUNTERS]  CudaCpp MEs      168.82
[madevent COUNTERS]  CudaCpp HEL      0.6287
--------------------------------------------------------------------------------
pp_dy3j.mad/cppavx2/output.txt
[GridPackCmd.launch] OVERALL TOTAL    251.9705 seconds
[madevent COUNTERS]  PROGRAM TOTAL    235.865
[madevent COUNTERS]  Fortran Overhead 165.362
[madevent COUNTERS]  CudaCpp MEs      70.1697
[madevent COUNTERS]  CudaCpp HEL      0.3335
--------------------------------------------------------------------------------
pp_dy3j.mad/cpp512y/output.txt
[GridPackCmd.launch] OVERALL TOTAL    246.8361 seconds
[madevent COUNTERS]  PROGRAM TOTAL    230.711
[madevent COUNTERS]  Fortran Overhead 165.532
[madevent COUNTERS]  CudaCpp MEs      64.8612
[madevent COUNTERS]  CudaCpp HEL      0.3183
--------------------------------------------------------------------------------
pp_dy3j.mad/cpp512z/output.txt
[GridPackCmd.launch] OVERALL TOTAL    267.4278 seconds
[madevent COUNTERS]  PROGRAM TOTAL    251.51
[madevent COUNTERS]  Fortran Overhead 165.54
[madevent COUNTERS]  CudaCpp MEs      85.562
[madevent COUNTERS]  CudaCpp HEL      0.4073
--------------------------------------------------------------------------------
pp_dy3j.mad/cuda/output.txt
[GridPackCmd.launch] OVERALL TOTAL    257.3246 seconds
[madevent COUNTERS]  PROGRAM TOTAL    215.547
[madevent COUNTERS]  Fortran Overhead 208.011
[madevent COUNTERS]  CudaCpp MEs      2.0941
[madevent COUNTERS]  CudaCpp HEL      5.4425
--------------------------------------------------------------------------------
```
This is to produce 100 events. Also to be understood/tested is whether I get the same lhe files in the different backends, the output seems to say 72 events in some cases and 100 elsewhere, so there is something fishy to be checked.

I will launch a test with 500 events overnight.

Anyway, just in case for the record, my tests are essentially done like this. From the installation above:
```
cd epochX/cudacpp/
# Generate the code (will take ~10 minutes)
./CODEGEN/generateAndCompare.sh pp_dy3j --mad
# Generate the multi-backend gridpack: compile/build all backends, prepare vegas grid using the cuda build (took me ~3h on a V100 + 4 CPU cores)
CUDACPP_RUNTIME_DISABLEFPE=1 ./tlau/lauX.sh -cuda pp_dy3j.mad -togridpack
# Run the gridpack for all backends (took me ~30 minutes for 100 events)
CUDACPP_RUNTIME_DISABLEFPE=1 ./tlau/lauX.sh -ALL pp_dy3j.mad -fromgridpack
```

Hope this helps, let me know in case of questions, cheers
Andrea

valassi commented:
PS just in case, the timers instrumentation is in my scripts, but you would not get it in running in the CMS mode. This is yet another PR upcoming at some point, even lower priority.

It would be interesting in any case to see if the general trend of how long it takes for 100 or 500 events is the same in your setup as what I observe. Otherwise we are doing something different.

Note: my tests and the times above are for one CPU core at a time. I think that in any case these are a good indication of what you would get as speedups in a realistic environment (you can always imagine using one CPU core at a time and filling each core with a different gridpack execution...).

valassi commented:
> @choij1589 (and @Saptaparna) if you really want to try this AS-IS no guarantees (and knowing that it will take 1-2 weeks before this gets merged in production - and knowing that the user interfaces will probably change), you can try my branch https://github.com/valassi/madgraph4gpu/tree/grid.

Hi again, just to say that there are clearly some issues, which may specific to this work or maybe more general
- the lhe events file produced look very different in fortran and cudacpp (#993), but I hav enot tested this on normal non-multi-backend builds either
- the cuda result is strange and has larger non-ME components than expected (#994)

So again, this is AS-IS. I will work on other things in the next few days, I am afraid this is all I can do for now...

choij1589 commented:
Hi @valassi, I will checked the PRs if it can be integrable (at least runnable regardless the outputs) to the CMS workflow. Thank you very much!

valassi commented:
Hi @choij1589 @Saptaparna just for info, I had to leave this running for 2-3 days but I have some results from DY+4j in this multi-backend gridpack
https://github.com/madgraph5/madgraph4gpu/blob/cf12395c0d662eaa9b05dba9950b0cd65e739fb2/epochX/cudacpp/tlau/fromgridpacks/pp_dy4j.mad/summary.txt
```
pp_dy4j.mad/fortran/output.txt (#events: 100)
[GridPackCmd.launch] OVERALL TOTAL    4824.4208 seconds
[madevent COUNTERS]  PROGRAM TOTAL    4671.95
[madevent COUNTERS]  Fortran Overhead 317.353
[madevent COUNTERS]  Fortran MEs      4354.6
--------------------------------------------------------------------------------
pp_dy4j.mad/cppnone/output.txt (#events: 44)
[GridPackCmd.launch] OVERALL TOTAL    5891.6592 seconds
[madevent COUNTERS]  PROGRAM TOTAL    5743.2
[madevent COUNTERS]  Fortran Overhead 315.785
[madevent COUNTERS]  CudaCpp MEs      5412.43
[madevent COUNTERS]  CudaCpp HEL      14.9908
--------------------------------------------------------------------------------
pp_dy4j.mad/cppsse4/output.txt (#events: 44)
[GridPackCmd.launch] OVERALL TOTAL    3129.8929 seconds
[madevent COUNTERS]  PROGRAM TOTAL    2977.71
[madevent COUNTERS]  Fortran Overhead 315.039
[madevent COUNTERS]  CudaCpp MEs      2655.23
[madevent COUNTERS]  CudaCpp HEL      7.446
--------------------------------------------------------------------------------
pp_dy4j.mad/cppavx2/output.txt (#events: 44)
[GridPackCmd.launch] OVERALL TOTAL    1589.4331 seconds
[madevent COUNTERS]  PROGRAM TOTAL    1438.59
[madevent COUNTERS]  Fortran Overhead 317.566
[madevent COUNTERS]  CudaCpp MEs      1117.72
[madevent COUNTERS]  CudaCpp HEL      3.2992
--------------------------------------------------------------------------------
pp_dy4j.mad/cpp512y/output.txt (#events: 44)
[GridPackCmd.launch] OVERALL TOTAL    1505.2116 seconds
[madevent COUNTERS]  PROGRAM TOTAL    1354.44
[madevent COUNTERS]  Fortran Overhead 316.385
[madevent COUNTERS]  CudaCpp MEs      1035.06
[madevent COUNTERS]  CudaCpp HEL      2.9977
--------------------------------------------------------------------------------
pp_dy4j.mad/cpp512z/output.txt (#events: 44)
[GridPackCmd.launch] OVERALL TOTAL    1585.8407 seconds
[madevent COUNTERS]  PROGRAM TOTAL    1440.68
[madevent COUNTERS]  Fortran Overhead 316.748
[madevent COUNTERS]  CudaCpp MEs      1120.05
[madevent COUNTERS]  CudaCpp HEL      3.884
--------------------------------------------------------------------------------
pp_dy4j.mad/cuda/output.txt (#events: 44)
[GridPackCmd.launch] OVERALL TOTAL    641.0242 seconds
[madevent COUNTERS]  PROGRAM TOTAL    458.101
[madevent COUNTERS]  Fortran Overhead 365.022
[madevent COUNTERS]  CudaCpp MEs      21.2839
[madevent COUNTERS]  CudaCpp HEL      71.7957
--------------------------------------------------------------------------------
pp_dy4j.mad/hip/output.txt
File not found: SKIP backend hip
----------------------------------------------------
```

In practice:
- AVX2 gives a factor ~4 for MEs and overall a factor ~3 speedup over Fortran (the non-ME part here is less important than in DY+3j, so the overall speedup is closer to the ME speedup)
- AVX512 is not better than AVX2 (but this is a Silver Intel CPU with 1 FMA unit; on a Gold Intel CPU I would expect a factor 2 ADDITIONAL for MEs, so maybe ~8 for MEs and ~6 overall faster than Fortran?
- CUDA is a factor ~200 better than Fortran for MEs and a factor ~8 better than Fortran overall (here the non-ME part like phase space sampling is dominant)

There are still problems that need to be understood however, in particular the LHE events file are not the same in fortran (which does give the required 100 events) and in cudacpp (which stops at 44 events). This needs to be investogated #993.

Anyway I think that the results are quite promising. I will make a slide for our dev meeting on Tuesday, let me know if you need anything for your meeting in CMS,. Thanks!

PS Ah NB the results above are on a 4-core node, but I am only using 1 CPU core in all tests from the gridpack, I will relaunch a test with 4 cores

valassi commented:
Ah the gridpack creation details are here
https://github.com/madgraph5/madgraph4gpu/pull/948/commits/338d5ec811ebc4ea9d31f0812718d803cfe1df40

In practice
- Around 1h for code generation - using 1 core
- Around 10h for the build of all backends (cuda, fortran, all cpp) - using `make -j` on 4 cores
- Around 1 day for creating the gridpack with CUDA i.e. optimising vegas grids - using 4 cores, so this is 4 days of CPU (the bottleneck is the CPU when the MEs are CUDA)

valassi commented:
Ah no actually run.sh does not have a nb_core option? So I guess this is it, single core generation results

choij1589 commented:
Hi @valassi , thank you very much for providing such a nice instruction. I have succeeded integrating multi-backend workflow into CMS gridpack_generation.sh, tested with DY+1j for simplicity. One thing I want to comment,

There is no hip related packages installed in current el8 image - but based on the logic for deciding possible backends, it should avoid compilation with hip backend:
[cudacpp.mk](https://github.com/choij1589/madgraph4gpu/blob/e786dee7fae099df40463532b7e0f99e81275bbd/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/madgraph/iolibs/template_files/gpu/cudacpp.mk#L1042-L1054)
But it launches bldhip. Currently I just modified the patch.common to go around this issue:
(patch.common)[https://github.com/choij1589/madgraph4gpu/blob/e786dee7fae099df40463532b7e0f99e81275bbd/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/MG5aMC_patches/PROD/patch.common#L273-L285]

Regarding the symlinks for event generation, I found it can be integrated smoothly in CMS's runcmsgrid.sh script and checked it runs with desired backends - many thanks! I will come back after some additional test with DY+3j and 4j, or inclusive one.

valassi commented:
> Hi @valassi , thank you very much for providing such a nice instruction. I have succeeded integrating multi-backend workflow into CMS gridpack_generation.sh, tested with DY+1j for simplicity. One thing I want to comment,
> 
> There is no hip related packages installed in current el8 image - but based on the logic for deciding possible backends, it should avoid compilation with hip backend: [cudacpp.mk](https://github.com/choij1589/madgraph4gpu/blob/e786dee7fae099df40463532b7e0f99e81275bbd/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/madgraph/iolibs/template_files/gpu/cudacpp.mk#L1042-L1054) But it launches bldhip. Currently I just modified the patch.common to go around this issue: [patch.common](https://github.com/choij1589/madgraph4gpu/blob/e786dee7fae099df40463532b7e0f99e81275bbd/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/MG5aMC_patches/PROD/patch.common#L273-L285)
> 
> Regarding the symlinks for event generation, I found it can be integrated smoothly in CMS's runcmsgrid.sh script and checked it runs with desired backends - many thanks! I will come back after some additional test with DY+3j and 4j, or inclusive one.

Thanks @choij1589 I am happy this helped, and I am happy that you could integrate it smoothly! :-)

Remember, many words of caution, this is very much WIP. And these strange issues #993 with numbers of events are worrying, for the functionality (but probably the performance numbers are reliable?).

Thanks about the hip comment. Maybe (as Olivier had mentioned at some point?) another option would be to allow a list of backends to build. Just thinking out loud for myself, the easiest here could be to provide a list of cudacpp.mk targets. Currently 'bldall' is 'bldhip bldcuda bldavxs' if cuda and hip exist, and 'bldavxs' itself' is 'bldnone bldsse4 bldavx2 bld512y bld512z'... so one way would be to just allow specifying a list like 'bldcuda bldnone bldsse4 bldavx2 bld512y bld512z'. I will keep this in mind.

I have some new results for DY+4j, I will post them in the next post below.

valassi commented:
Voila these are my latest results from this weekend, with 500 DY+4jets (previously I had 100 DY+4j events)

https://github.com/madgraph5/madgraph4gpu/blob/d3da4603250b625073f6591c4bd69dea914a3c51/epochX/cudacpp/tlau/fromgridpacks/pp_dy4j.mad/summary.txt
```
pp_dy4j.mad/fortran/output.txt (#events: 81)
[GridPackCmd.launch] OVERALL TOTAL    21707.6095 seconds
[madevent COUNTERS]  PROGRAM TOTAL    21546.1
[madevent COUNTERS]  Fortran Overhead 1579.09
[madevent COUNTERS]  Fortran MEs      19967
--------------------------------------------------------------------------------
pp_dy4j.mad/cppnone/output.txt (#events: 195)
[GridPackCmd.launch] OVERALL TOTAL    26745.1639 seconds
[madevent COUNTERS]  PROGRAM TOTAL    26584.9
[madevent COUNTERS]  Fortran Overhead 1608.51
[madevent COUNTERS]  CudaCpp MEs      24910.4
[madevent COUNTERS]  CudaCpp HEL      66.0341
--------------------------------------------------------------------------------
pp_dy4j.mad/cppsse4/output.txt (#events: 195)
[GridPackCmd.launch] OVERALL TOTAL    14398.4664 seconds
[madevent COUNTERS]  PROGRAM TOTAL    14231.3
[madevent COUNTERS]  Fortran Overhead 1647.03
[madevent COUNTERS]  CudaCpp MEs      12550.6
[madevent COUNTERS]  CudaCpp HEL      33.7035
--------------------------------------------------------------------------------
pp_dy4j.mad/cppavx2/output.txt (#events: 195)
[GridPackCmd.launch] OVERALL TOTAL    7335.2356 seconds
[madevent COUNTERS]  PROGRAM TOTAL    7114.43
[madevent COUNTERS]  Fortran Overhead 1683.7
[madevent COUNTERS]  CudaCpp MEs      5415.48
[madevent COUNTERS]  CudaCpp HEL      15.2596
--------------------------------------------------------------------------------
pp_dy4j.mad/cpp512y/output.txt (#events: 195)
[GridPackCmd.launch] OVERALL TOTAL    6831.8971 seconds
[madevent COUNTERS]  PROGRAM TOTAL    6649.98
[madevent COUNTERS]  Fortran Overhead 1669.94
[madevent COUNTERS]  CudaCpp MEs      4966.24
[madevent COUNTERS]  CudaCpp HEL      13.8066
--------------------------------------------------------------------------------
pp_dy4j.mad/cpp512z/output.txt (#events: 195)
[GridPackCmd.launch] OVERALL TOTAL    7136.2962 seconds
[madevent COUNTERS]  PROGRAM TOTAL    6958.96
[madevent COUNTERS]  Fortran Overhead 1636.28
[madevent COUNTERS]  CudaCpp MEs      5305.14
[madevent COUNTERS]  CudaCpp HEL      17.5447
--------------------------------------------------------------------------------
pp_dy4j.mad/cuda/output.txt (#events: 195)
[GridPackCmd.launch] OVERALL TOTAL    2523.7488 seconds
[madevent COUNTERS]  PROGRAM TOTAL    2234.93
[madevent COUNTERS]  Fortran Overhead 1820.36
[madevent COUNTERS]  CudaCpp MEs      97.9622
[madevent COUNTERS]  CudaCpp HEL      316.613
--------------------------------------------------------------------------------
```

Essentially for DY+4j I get exactly the same trend with 500 events that I saw with 100 events
- AVX2 gives a factor 4 for MEs and overall a factor 3 speedup over Fortran
- AVX512 is not better than AVX2 (on a Silver Intel CPU; on a Gold Intel CPU I would expect maybe 8 for MEs and 6 overall faster than Fortran?)
- CUDA is a factor 200 better than Fortran for MEs and a factor 8 better than Fortran overall

A few issues of concern are even more obvious
- #993 the numbers of events and the LHE files are different in fortran vs cudacpp
- #994 the Fortran non-ME part in cuda (1820s) is larger than in fortran or cpp (1600-1650s)
- #999 the helicity-filtering component is strangely very high in cuda (316s)
- #1000 the python/bash component is higher in cuda (300s) than in fortran or cpp (150-180s, which is still high and to be understood)

choij1589 commented:
@valassi Good to see overall speedups:) Despite the issues, I am testing with DY+4j/01234j, TT+3j/0123j - I will come back as soon as I have the results.

choij1589 commented:
> There are still problems that need to be understood however, in particular the LHE events file are not the same in fortran (which does give the required 100 events) and in cudacpp (which stops at 44 events). This needs to be investogated #993.

Another thing I've noticed while comparing timing for event generation with TT+0123j / 50k events, 8 times in parallel,
                    Single Backend vs Multi-Backend
FORTRAN: 1228 \pm ￼89m vs 1167 \pm ￼ 92m 
CPPAVX2:  638 ￼ \pm 92m vs 599 ￼\pm 17m 
CUDA:        226 ￼ \pm 22m vs 120 ￼\pm 8m

So CUDA with multi-backend consumes only half of the timing compared to single-backend CUDA gridpack. This might related to the difference no. of events produced for each backend?

choij1589 commented:
Hi @valassi, yes my multi-backend gridpack was indeed built with cuda. Sorry I misunderstood that CUDA generates LESS events than required events with this PR...(e.g. if I require 100 events to be generates it gives 44 events) but I checked both event generation with single cuda-backend and multi-backend give the same no. of 50k events. 
One thing I am suspicious now is - was there further optimization for CUDA between 19th August and these days? My single cuda backend gridpack was produced in the middle of the August.

valassi commented:
Thanks @choij1589 for the feedback. I am not sure about your questions, I will need to have a more detailed look if there were any changes that may impact what you see. When I have time I will do that and keep you posted.

valassi commented:
I have merged v1.00.00 master (and also #1014) into this branch and regenerated all code.

valassi commented:
> I have merged v1.00.00 master (and also https://github.com/madgraph5/madgraph4gpu/pull/1014) into this branch and regenerated all code.

Same thing but with a better history to keep track of some things

valassi commented:
And now including also upstream/master with the AMD and v1.00.01 patches #1014 and #1012

valassi commented:
I removed the link to #947 (this was fixed in https://github.com/mg5amcnlo/mg5amcnlo/pull/125, submodule was updated in PR https://github.com/madgraph5/madgraph4gpu/pull/949)

valassi commented:
Hi @oliviermattelaer as discussed via email.

This is N=3 in the pipeline I would like to merge.

I changed this to target the N=2 makefiles branch for easier reviewing, but I would then merge it against master once approved.

Can you please tell me if this looks ok? Thanks
Andrea

valassi commented:
(Note: this also addresses #700 but I did not manage to link it yet)

oliviermattelaer commented:
# Point by Point review

## multi backend gridpacks (cudacpp_bldall runcard)

No problem on that one, but maybe why including it by default in the run_card?
If you have a comment:
```
 WARNING! The following cudacpp runcards are experimental! Users should normally change only the cudacpp_backend card ***
```
Would it make more sense to not display them at all (or in a independent hidden block "simd_advance"?)
-> to be discussed

## add cudacpp_helinl and cudacpp_hrdcod runcards for HELINL=1 and HRDCOD=1 modes

 - Same comment as above
 - comment on the default_setup() need to be changed from a question to an assertion
 - maybe good to add the comment field for those?
 
## rename the FPTYPE runcard as cudacpp_fptype

This is delicate since this is going to break retro-compatibility of the code which we have to preserve.
So we either have to not do this change (which is ok) or do it but add a mechanism preserving retro-compatibility.
-> to be discussed

## instrument gridpacks with time profiling of the python/bash orchestrator (this is only in my "tlau" tests for the moment, eventually it can be moved in mg5amcnlo, but for the moment it is there so I can use it and further develop it)

I do not see any change in the codegen directory so I guess that you edit manually a .tmad directory, which is fine since those directory needs anyway to be remove from the repo.

# Global Comment

Touching the patch, should be avoid as much as possible, do you want to add new handle in the template?
Or do you want to do it? Would be nice to avoid to touch that makefile in such a way...

valassi commented:
Hi Olivier thanks for the comments :-)

Maybe easiest to have a short zoom when you have time tomorrow or another time.

In short

- As you prefer for the visibility of the cards. I find it nice to have them in the cards, because this way it's easy to add a comments for users about which options are available. But I do not know how users use runcards enough to judge. Maybe there is some other way to document which cards are available and which options are available for each card.

- As you prefer for floating_type, ok for me to keep it that way. I thought cudacpp_fptype was more consistent, but I really do not care much. I had not thought that backward compatibility was an issue.

- Thanks for accepting the concept of multi backend gridpacks and of python profiling. 

- Your comment about "deault_setup" I have not understood. To be discussed.

- About "Touching the patch, should be avoid as much as possible, do you want to add new handle in the template?", I guess you mean that we should go in the direction of what I did for driver.f in https://github.com/mg5amcnlo/mg5amcnlo/pull/148. My answer is yes, eventually I'd like to do that, but I have not done it here. For the moment the only thing I am trying to do is to NOT INCREASE the stuff in patch.P1 and patch.common. Thanks to various changes you did we reduced this to three files (SubProcesses/makefile, driver.f, matrix1.f), and with what I did fro driver.f I plan to also get rid of driver.f. Then yes we should also get rid of the patch for makefile. I would just do this later... but if you prefer, I will try to get rid of these three first thing (probably after CHEP).

Apart from this:
- I have merged the two base PRs
- So I have updated this PR accordingly
- And I changed its base to master as now it is the lowest level in my "pipeline"

Lets dscuss the other points above by zoome (or just reply here in the meantime), tomorrow.

Thanks! :-)
Andrea
